pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,249 @@
+		}, func() float64 {
+			if currentStateFn() == circuitbreaker.HalfOpenState {
+				return 1
+			}
+			return 0
+		}),
+		circuitBreakerClosedStateGauge: promauto.With(r).NewGaugeFunc(prometheus.GaugeOpts{
+			Name:        circuitBreakerCurrentStateGaugeName,
+			Help:        circuitBreakerCurrentStateGaugeHelp,
+			ConstLabels: map[string]string{circuitBreakerCurrentStateGaugeLabel: circuitbreaker.ClosedState.String()},
+		}, func() float64 {
+			if currentStateFn() == circuitbreaker.ClosedState {
+				return 1
+			}
+			return 0
+		}),
+		circuitBreakerTransitions: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_circuit_breaker_transitions_total",
+			Help: "Number of times the circuit breaker has entered a state.",
+		}, []string{"ingester", "state"}),
+		circuitBreakerResults: promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingester_circuit_breaker_results_total",
+			Help: "Results of executing requests via the circuit breaker.",
+		}, []string{"ingester", "result"}),
+	}
+}
+
+type CircuitBreakerConfig struct {
+	Enabled                   bool          `yaml:"enabled" category:"experimental"`
+	FailureThreshold          uint          `yaml:"failure_threshold" category:"experimental"`
+	FailureExecutionThreshold uint          `yaml:"failure_execution_threshold" category:"experimental"`
+	ThresholdingPeriod        time.Duration `yaml:"thresholding_period" category:"experimental"`
+	CooldownPeriod            time.Duration `yaml:"cooldown_period" category:"experimental"`
+	InitialDelay              time.Duration `yaml:"initial_delay" category:"experimental"`
+	PushTimeout               time.Duration `yaml:"push_timeout" category:"experiment"`
+	testModeEnabled           bool          `yaml:"-"`
+}
+
+func (cfg *CircuitBreakerConfig) RegisterFlags(f *flag.FlagSet) {
+	prefix := "ingester.circuit-breaker."
+	f.BoolVar(&cfg.Enabled, prefix+"enabled", false, "Enable circuit breaking when making requests to ingesters")
+	f.UintVar(&cfg.FailureThreshold, prefix+"failure-threshold", 10, "Max percentage of requests that can fail over period before the circuit breaker opens")
+	f.UintVar(&cfg.FailureExecutionThreshold, prefix+"failure-execution-threshold", 100, "How many requests must have been executed in period for the circuit breaker to be eligible to open for the rate of failures")
+	f.DurationVar(&cfg.ThresholdingPeriod, prefix+"thresholding-period", time.Minute, "Moving window of time that the percentage of failed requests is computed over")
+	f.DurationVar(&cfg.CooldownPeriod, prefix+"cooldown-period", 10*time.Second, "How long the circuit breaker will stay in the open state before allowing some requests")
+	f.DurationVar(&cfg.InitialDelay, prefix+"initial-delay", 0, "How long the circuit breaker should wait between creation and starting up. During that time both failures and successes will not be counted.")
```

Comment:
What's the thinking behind InitialDelay - does this seem needed? Is there a risk if it's too short?

Reply:
The idea behind `InitialDelay` is to allow circuit breakers to open only after the initial delay has passed. By default this configuration is set to 0, meaning that the circuit breakers can open as soon as the conditions for that are met. But, if we want to disable circuit breaker opening during a weekly rollout for example, we could set this value to 5 minutes or something like that, so that no circuit breaker will open until all the ingester from all the zones have completed the rollout.